### ISSUE TITLE
Different keyboard layouts for individual windows

### ISSUE THREAD

**Issue by liminspace** [elementary/gala#180]
Hi there!
It's very important to have different layouts for each windows if you use more than one layout.
But it is impossible in Elementary OS.
Is there some solution how to fix it?
Thanks.

ps: Elementary OS can be the first candidate to move from Ubuntu with Unity, but it must have at least base functionality like above.

**Comment by peteruithoven** [comment]
Could you explain why it's important? This sounds very subjective when you don't add some basic reasoning.

**Comment by liminspace** [comment]
At least because it is not comfortable when you use several programs and must change layout every time changing window. After Ubuntu, Windows, MacOS or other OS, Elementary OS is very uncomfortable for its woody layouts.
If you use more that one layout you must understand me without additional explanation.

**Comment by peteruithoven** [comment]
>  If you use more that one layout you must understand me without additional explanation.

Clearly I don't, which is why I'm asking this. Why are people using multiple layouts? I don't represent elementary OS, but I'd like to see some reasoning and clarification on the importance of this feature request.

**Comment by liminspace** [comment]
Imagine you have a messenger, an email client, several terminals and a code editor. You use some A language in messenger, B in email client and C in terminals and the code editor. If OS doesn't support multi-window layouts, you must every time change the layout to write text in the right language and check if current layout is correct. If OS supports this feature, you can just change window without thinking about layout and spending time for checking and switching it.

**Comment by oleg104** [comment]
This option is very important for all people who are from not english-speaking country like me and Ihor. We always have at least two input languages. Most of us has three input languages/loyouts. We use different languages for Internet browsing, messaging, programming. Please add option "Different keyboard layouts for individual windows" to Elementary OS. I saw that this option was present earlier like here https://goo.gl/images/75qiGc

**Comment by SunDrop** [comment]
I do agree that this option is very important!
Can you show the way to return this option?

**Comment by liminspace** [comment]
I just moved to Ubuntu 18.04. After some tuning it is quite comfortable and stable.

**Comment by gasinvein** [comment]
Lack of this feature is probably a showstopper for anybody whose native language isn't english.

**Comment by MegOBonus** [comment]
Any updates?

**Comment by peteruithoven** [comment]
Please stop asking for updates. Maybe add a bounty or try to put in some time instead?

**Comment by axelprox** [comment]
Maybe someone already solve this? Or AppCenter has app to deal with it?

**Comment by Dirli** [comment]
Maybe someone decided that)) If there is a wish to test, welcome
https://github.com/Dirli/gala-layoutpw-plugin

**Comment by endorama** [comment]
Hello, I really need this! Switching back and forth from different layout is a pity, and I really love the polished touch Elementary has. Thus I'm willing to help, but I'm new to Vala and Gala. 

@peteruithoven do you think this is something a "newcomer" can tackle? I'm fairly experienced in other languages, but I mainly coded cli tools and web (both frontend and backend) apps.  
If yes I'll gladly have a try.

**Comment by ochornolutskyi** [comment]
> > Здравствуйте, я хотел бы видеть это поддерживается! Жаль, что переключаться с другой компоновки на другую - мне очень нравится, что элементарное прикосновение есть у Elementary. Таким образом, я готов помочь, но я новичок в Vala и Gala (и в программировании GTK в целом).
> > @peteruithoven Как вы думаете, это то, что новичок может решить? Я довольно опытен в других языках, но я в основном программировал инструменты для работы с клиентами и веб-приложения (как веб-, так и серверные). 
> > Если да, я с удовольствием попробую.
> 
> Вы можете попробовать этот
> [https://github.com/Dirli/gala-layoutpw-plugin,](https://github.com/Dirli/gala-layoutpw-plugin)
> и этот патч для включения / выключения (0.3.1)
> [https://github.com/Dirli/gpde/blob/master/pantheon- основание / распределительный щит-модуль клавиатуры / файлы / 2.3.5-layout_pw_switch.patch](https://github.com/Dirli/gpde/blob/master/pantheon-base/switchboard-plug-keyboard/files/2.3.5-layout_pw_switch.patch)


I tried to install your plugin but it doesn't work for me. Installed according to the installation instructions. Maybe I did something wrong?

**Comment by Dirli** [comment]
> > > Здравствуйте, я хотел бы видеть это поддерживается! Жаль, что переключаться с другой компоновки на другую - мне очень нравится, что элементарное прикосновение есть у Elementary. Таким образом, я готов помочь, но я новичок в Vala и Gala (и в программировании GTK в целом).
> > > @peteruithoven Как вы думаете, это то, что новичок может решить? Я довольно опытен в других языках, но я в основном программировал инструменты для работы с клиентами и веб-приложения (как веб-, так и серверные).
> > > Если да, я с удовольствием попробую.
> > 
> > 
> > Вы можете попробовать этот
> > [https://github.com/Dirli/gala-layoutpw-plugin,](https://github.com/Dirli/gala-layoutpw-plugin)
> > и этот патч для включения / выключения (0.3.1)
> > [https://github.com/Dirli/gpde/blob/master/pantheon- основание / распределительный щит-модуль клавиатуры / файлы / 2.3.5-layout_pw_switch.patch](https://github.com/Dirli/gpde/blob/master/pantheon-base/switchboard-plug-keyboard/files/2.3.5-layout_pw_switch.patch)
> 
> I tried to install your plugin but it doesn't work for me. Installed according to the installation instructions. Maybe I did something wrong?

Use 0.3.0 version.
0,3,1 is disabled by default and must be enabled. patch:
https://github.com/Dirli/gpde/blob/master/pantheon-base/switchboard-plug-keyboard/files/2.3.5-layout_pw_switch.patch

**Comment by saintzyo** [comment]
+1 for this feature request. All "adult" OSs have this feature and even more: per app, per window for each app. This is like adding another input language. it's the only thing that stops me from migrating from Ubuntu to Elementary. And the slogan "User Friendly. And Keyboard Friendly." sounds a little bit defective.

**Comment by ochornolutskyi** [comment]
> +1 for this feature request. All "adult" OSs have this feature and even more: per app, per window for each app. This is like adding another input language. it's the only thing that stops me from migrating from Ubuntu to Elementary. And the slogan "User Friendly. And Keyboard Friendly." sounds a little bit defective.

Hello, my friend. Sorry for my Ukrainian English, but this plugin "https://github.com/Dirli/gala-layoutpw-plugin" really working for me. After installing it plugin as recommended,  please, install dconf-editor. Then you must change "per window" on "true" for "/org/gnome/desktop/input-sources/per-window" if you use GTK. Or try to run in console "dconf write /org/gnome/desktop/input-sources/per-window true". It works for me after rebooting and I am so happy.

**Comment by saintzyo** [comment]
> > +1 for this feature request. All "adult" OSs have this feature and even more: per app, per window for each app. This is like adding another input language. it's the only thing that stops me from migrating from Ubuntu to Elementary. And the slogan "User Friendly. And Keyboard Friendly." sounds a little bit defective.
> 
> Hello, my friend. Sorry for my Ukrainian English, but this plugin "https://github.com/Dirli/gala-layoutpw-plugin" really works for me. After installing it plugin as recommended, please, install dconf-editor. Then you must change "per window" on "true" for "/org/gnome/desktop/input-sources/per-window" if you use GTK. Or try to run in console "dconf write /org/gnome/desktop/input-sources/per-window true". It works for me after rebooting and I am so happy.

Yup, it maybe works. But, I want to have this feature out from the box. I don't want to take any extra steps. And if Elementary is positioning herself as "The fast, open, and privacy-respecting replacement for Windows and macOS", so where is this?)
Don't worry, I'm also from Ukraine!)

**Comment by adovbos** [comment]
+1 for this feature. And +1 to Ukrainian users)))

**Comment by mrduke243** [comment]
Hello, 

I am trying the above, I seem to be having an issue with the instructions shown for the plug in or probably i am doing something wrong, I've installed the dependencies needed:

![image](https://user-images.githubusercontent.com/63676079/79271538-a66d2e00-7ea8-11ea-9ff5-20b69a99f9a4.png)

Then i am supposed to write this command:

' meson build --prefix=/usr //debian,ubuntu --libdir=/usr/lib/x86_64-linux-gnu '

Here is my outcome:
' usage: meson [-h]
             {setup,configure,install,introspect,init,test,wrap,subprojects,help}
             ...
meson: error: unrecognized arguments: //debian,ubuntu '

Now, to my understanding - meson cannot be used with " build ", also my arguments are wrong - i tried filling in ' elementary ' instead of ' debian, ubuntu ' - I am still getting the same outcome though. (only difference is that i get that ' elementary ' is an invalid argument.

Could someone help or point to me what i am doing wrong here?


Thanks!

**Comment by adovbos** [comment]
Just delete //Debian, Ubuntu... "//" Usually means " this is comment"

**Comment by mrduke243** [comment]
Command : ' meson build --prefix=/usr --libdir=/usr/lib/x86_64-linux-gnu '

I believe i am missing something here, i also tried removing, i am getting this message though:

ERROR: Neither directory contains a build file meson.build.


Note: I downloaded the plug in and extracted on my Downloads folder so probably it's a matter of 'wrong' directory?

**Comment by ochornolutskyi** [comment]
> Command : ' meson build --prefix=/usr --libdir=/usr/lib/x86_64-linux-gnu '
> 
> I believe i am missing something here, i also tried removing, i am getting this message though:
> 
> ERROR: Neither directory contains a build file meson.build.
> 
> Note: I downloaded the plug in and extracted on my Downloads folder so probably it's a matter of 'wrong' directory?

please make sure you go into the plugin folder in the terminal.
like this: 
cd ~/your_folders_name/gala-layoutpw-plugin/
and then try installing it one more time

**Comment by mrduke243** [comment]
Thanks, that worked while I was logged in. I restarted my laptop and now I cannot go past the password logo in screen.


It accepts the pw, and when asks for it again.. Non stop?


Any thoughts, apart from clean re install?

**Comment by adovbos** [comment]
> Thanks, that worked while I was logged in. I restarted my laptop and now I cannot go past the password logo in screen.
> 
> It accepts the pw, and when asks for it again.. Non stop?
> 
> Any thoughts, apart from clean re install?

sounds more like wingpanel or pantheon error. Try to go on other tty using ctrl+alt+f1-12 and login there, and see sys logs for errors, or remove this plugin using uninstall command

**Comment by ochornolutskyi** [comment]
> Thanks, that worked while I was logged in. I restarted my laptop and now I cannot go past the password logo in screen.
> 
> It accepts the pw, and when asks for it again.. Non stop?
> 
> Any thoughts, apart from clean re install?

I don't know some things in this plugin, I am not the creator of it. I think the plugin works non stop. it's ok for me because my password includes only figures. I think you can use the live cd or do some things that @adovbos tell above, and remove the plugin from: `cd /usr/lib/x86_64-linux-gnu/gala/plugins/ ` and then `sudo rm libgala-layoutpw.so`. Did you try to change your layout on lockdown screen?

**Comment by mrduke243** [comment]
Doesn't work sadly.

Thanks though, will clean install instead. It's faster since I just installed a while ago

**Comment by jeremypw** [comment]
I had a similar problem a while back and I just logged in using a terminal (press e.g. `<Ctrl><Alt>F5` at the log in screen) and updated the system. You could also create a new user for yourself from the terminal if that does not fix it.  I am always building and installing proposed branches so I assumed that had messed something up so I did not report it as in issue at the time.

**Comment by mrduke243** [comment]
I tried updating, I didn't fix the issue...

I found a "solution" though, since I'm the only user of my laptop I just removed the pw log in so it doesn't mess up my log in credentials.

It's not a solution per se, but for the sake of using this 😅

**Comment by Dirli** [comment]
It is very strange that you are trying to find a solution to the problem here, and not from the author of this plugin)))

PS Yes, there was an error in the master branch (or rather not mindfulness). Everything is fixed (including the description).

**Comment by mrduke243** [comment]
I saw the post and figured it was worth a try to ask in case someone has a similar problem, usually it's the fastest way to go 😁

**Comment by tutanotaaa** [comment]
Two years later and no option for different layouts in apps? Are you serious?

**Comment by jeremypw** [comment]
@Dirli:  Have you thought about submitting a PR to elementary incorporating your plugin into the master branch?

**Comment by Dirli** [comment]
@jeremypw  Yes, that would be a good idea. But the position of the main developers is that they are not interested in what they do not need. I'm afraid it will be a waste of time, or the PR will hang for years. So I'll leave it at that. If someone does this, no problem.

**Comment by megatux** [comment]
I also would like to see this plugin merged, unless there are _good reasons_ not to do it, like UX complexity, code quality, open issues, etc.

Sorry not to know who are the "main developers" on this thread. Could you list the reasons for not to apply this solution or the steps needed for move forward with it?

**Comment by jeremypw** [comment]
I don't work on Gala.  I am not sure there is an "official" maintainer but people recently making substantive commits include @tintou, @donadigo and @davidmhewitt .

**Comment by A1Gard** [comment]
@Dirli 
Thanks for your lib it's worked perfectly. 
But I think better is easier approach. 

Spasibo
